### DISCUSSION TITLE
App 组件使用了 a 标签的全局样式，造成了样式污染

### DISCUSSION THREAD

**Discussion by Minf97** [ant-design/ant-design#50426] [General]
在我的 NextJs@14 项目中，今天意外发现我的 a 标签颜色都错误了。检查发现多了这么一段全局样式

![image](https://github.com/user-attachments/assets/107b9d64-82c9-48c9-a4c2-67ee697bf349)

可以立刻判断出这是来自 antd 的全局样式，但具体在哪里并不清楚。

我甚至开始怀疑 "这个页面没有用 antd 组件也会引入样式，难道不应该 tree shaking 吗"

我花了很长时间定位这段全局样式的位置，最终在我一次一次回滚 commit 中，定位到我在某一次 commit 里使用了 `<App></App>` ，原因是我想全局统一 `Modal`

回归正题，能否把这段样式污染去除掉？

**Comment by Wxh16144** [comment]
> 我甚至开始怀疑 "这个页面没有用 antd 组件也会引入样式，难道不应该 tree shaking 吗"

antd 所有组件下的 a 标签都自带样式，不知道为什么这么做，或许在 v6 回考虑去掉这个操作。

开发者只能覆盖样式， 或者全局修改 `theme.token.colorLink` token

与 a 标签相关的 token 应该有 `colorLink` 、`colorLinkActive` 、 `colorLinkHover`

ref：https://github.com/ant-design/ant-design/blob/b34d20860c9b58144a6b8e053944b865d75062e8/components/style/index.tsx#L66-L100

**Comment by arvinxx** [comment]
App 下 的 a 标签加上样式是合理的设计啊，相当于在 App 作用域下的 link 都是符合 Ant Design 设计规范定义的样式。否则不加的话就会是这样：
 
<img width="1378" alt="image" src="https://github.com/user-attachments/assets/4824c132-1dd4-4a1c-96ba-42eb87de738e"> 

https://stackblitz.com/edit/vitejs-vite-svwgbd?file=src%2Findex.css,src%2FApp.tsx,src%2FApp.css,src%2Fmain.tsx&terminal=dev

我觉得这不属于样式污染的范畴。如果你不满意 antd 默认的 link 样式，加个 global.css 全局覆写 a 样式也完全 OK 的吧。

**Reply by Minf97** [reply]
那么在 global.css 全局怎么覆写呢？

事实上我项目里本身就有 global.css，而不加  `!important` 优先级无法高过 antd。但加 `!important` 同样不是我想要的。

@example
期望结果：a 标签跟随父盒子文字颜色为 `test-gray-50`
实际结果：a 标签优先是 `global.css !important`
```html
<div className="text-gray-50">
     <a> 测试文案 </a>
</div>
```

**Reply by Wxh16144** [reply]
那就从 css 选择器上加权重 `* a {}`

**Reply by Minf97** [reply]
> 那就从 css 选择器上加权重 `* a {}`

很遗憾是不行的，我刚才尝试在权重上寻找高于 `:where(css-dev-only)` 且低于 `类选择器`(覆盖全局样式又不影响原本的样式)，前者的权重特异性是 (0,0,1)，后者的权重特异性是(0,1,0)。没有介乎中间的值

我最后的做法还是去掉 App 了，此贴结束

**Reply by afc163** [reply]
试试 `<App component={false}>`？

**Reply by arvinxx** [reply]
@Minf97 https://stackblitz.com/edit/stackblitz-starters-rfcte5?description=The%20React%20framework%20for%20production&file=app%2Fpage.tsx,app%2Fglobals.css&title=Next.js%20Starter

没理解为什么全局不行，我试下来直接这样就好了： 

```css
a {
  color: inherit;
}

a:hover {
  color: inherit;
}
```

如果这样不行，大概率可能是你的实现中存在其他什么干扰项。

<img width="1008" alt="image" src="https://github.com/user-attachments/assets/a731592f-5229-44d0-8a39-12bcb011f2fa">


<img width="1105" alt="image" src="https://github.com/user-attachments/assets/8c4d0393-e201-4d18-8ead-11ab5d892caa">

**Reply by Minf97** [reply]
> 没理解为什么全局不行，我试下来直接这样就好了：

我改写了一下 demo，添加了` AntdRestry`

https://stackblitz.com/edit/stackblitz-starters-fwa52l?description=The%20React%20framework%20for%20production&file=package.json,app%2Flayout.tsx,app%2Fpage.tsx&title=Next.js%20Starter

**Reply by Minf97** [reply]
> 试试 `<App component={false}>`？

谢谢你！这段代码解决了我的问题！

**Reply by arvinxx** [reply]
> 我改写了一下 demo，添加了 AntdRestry

原因是 AntdRegistry 组件将收集来的样式插在了 head 的最下方，按照同优先级样式，下面的会先生效：

<img width="1016" alt="image" src="https://github.com/user-attachments/assets/f3658a7f-0748-48b9-8c23-765df5956992">

如果换一下顺序，可以看到就正常了：

<img width="1041" alt="image" src="https://github.com/user-attachments/assets/72a58a47-3c6c-4081-9117-7911fa6a0a59">

@MadCcc 我在想是不是 AntdRegistry 应该先找一下 `<link rel="stylesheet" href="/_next/static` 这样的 link标签，然后插在它的前面呢？不知道 useServerInsertedHTML 有没有办法控制是 append 还是 prepend。

----

不过既然你引入 StyleProvider 了（AntdRegistry 的依赖），还有一种更简单的解法是使用 `layer` 字段直接将 antd 的所有样式直接降权，这样可以都不需要写 `color: inherit;` 。你的任何 tailwind class 都可以覆写掉antd原有的样式。

<img width="1087" alt="image" src="https://github.com/user-attachments/assets/19e49a3d-a70b-448d-b0ca-7d17f1a9eb7c">

<img width="1048" alt="image" src="https://github.com/user-attachments/assets/20cf59a8-ca83-4c70-8bca-e6b6c644739d">

实现上就在 AntdRegistry 加个 layer 就好。 （https://ant-design.antgroup.com/docs/react/compatible-style-cn#layer-%E9%99%8D%E6%9D%83 ）

<img width="1436" alt="image" src="https://github.com/user-attachments/assets/089b26bf-7b9b-4f4c-8f10-7c6b11751b37">

不过 layer 也有代价，Chrome 99 版本才开始支持（https://developer.mozilla.org/en-US/docs/Web/CSS/@layer ）

**Reply by Minf97** [reply]
> 原因是 AntdRegistry 组件将收集来的样式插在了 head 的最下方，按照同优先级样式，下面的会先生效：

> 不过 layer 也有代价，Chrome 99 版本才开始支持（https://developer.mozilla.org/en-US/docs/Web/CSS/@layer ）

layer 的版本限制太高，需要 Chrome 99 2022 年后的版本才可支持（https://caniuse.com/?search=%40layer ），这对我来说是不可接受的，目前我已采用楼上 afc 大佬的`<App component={false}></App>`

感谢大佬您的认真积极回复！orz



---

以上关于 layer 的版本要求，或许可以补充到文档（https://ant-design.antgroup.com/docs/react/compatible-style-cn#layer-%E9%99%8D%E6%9D%83 ）

**Reply by arvinxx** [reply]
@Minf97 不妨来提个 PR 😄
